### ISSUE TITLE
Repeatedly calling bfgs_minimize on Jax creates memory leak

### ISSUE THREAD

**Issue by simitii** [tensorflow/probability#1641]
When I call bfgs_minimize as follow repeatedly (in a for loop) on Jax, it creates memory leak (memory usage increases after each iteration):

`loss = lambda w: objective_function(w, z, y, F, a)`
`optim_results = tfp.optimizer.bfgs_minimize(value_and_grad(loss), initial_position=w0.copy())`

Jax Version: 0.3.14
Tensorflow-Probability Version: 0.17.0-dev20220808

**Comment by simitii** [comment]
I found out it was about the caching mechanism of Jax.

https://github.com/google/jax/issues/10828#issuecomment-1272809353 solves it.
